src/natcap/invest/datastack.py

```diff
@@ -311,77 +192,217 @@ def build_datastack_archive(args, model_name, datastack_path):
+                                column_value == ''):
+                            # The table cell is blank, so skip it.
+                            # We can't compare nan values directly in a way
+                            # that also works for strings, so skip it.
+                            continue
+
+                        source_filepath = None
+                        for possible_filepath in (
+                                column_value,
+                                os.path.join(csv_source_dir, column_value)):
+                            if os.path.exists(possible_filepath):
+                                source_filepath = possible_filepath
+                                break
+
+                        # If we didn't end up finding a valid source filepath
+                        # for the field value, assume it's supposed to be that
+                        # way and leave it alone.
+                        if not source_filepath:
+                            continue
+
+                        try:
+                            # This path is already relative to the data
+                            # directory
+                            target_filepath = files_found[source_filepath]
+                        except KeyError:
+                            basename = os.path.splitext(
+                                os.path.basename(source_filepath))[0]
+                            target_dir = os.path.join(
+                                contained_files_dir,
+                                f'{row_index}_{basename}')
+                            target_filepath = _copy_spatial_files(
+                                source_filepath, target_dir)
+                            target_filepath = os.path.relpath(
+                                target_filepath, data_dir)
+
+                        LOGGER.debug(
+                            'Spatial file in CSV copied from '
+                            f'{source_filepath} --> {target_filepath}')
+                        dataframe.at[
+                            row_index, spatial_column_name] = target_filepath
+                        files_found[source_filepath] = target_filepath
+
+                LOGGER.debug(
+                    f'Rewritten spatial CSV written to {target_csv_path}')
+                dataframe.to_csv(target_csv_path)
+
+            target_arg_value = _relpath(target_csv_path)
+            files_found[source_path] = target_arg_value
+
+        elif input_type == 'file':
+            target_filepath = os.path.join(
+                data_dir, f'{key}_{os.path.basename(source_path)}')
+            shutil.copyfile(source_path, target_filepath)
+            LOGGER.debug(
+                f'File copied from {source_path} --> {target_filepath}')
+            target_arg_value = _relpath(target_filepath)
+            files_found[source_path] = target_arg_value
+
+        elif input_type == 'directory':
```

Comment:
Directories are an interesting case. I wonder if you gave some thought to these issues:

1. Do we want to include the `workspace_dir` and it's contents? It could be very useful to do so. But it might be a little unexpected if the user is expecting to just archive the input data.
2. Could there be directories that contain a lot of other files besides the ones the model will use? E.g. seasonal water yield `et0_dir` looks like it does some pattern matching to find the files within the dir. Maybe it's okay though, to archive extra files. It's unlikely to happen in practice.

Reply:
Ooo good point about the `workspace_dir`.  I just added an explicit exclusion there, lest we one day end up with a massive datastack that has a user's complete workspace in there!  1d85c86d8

> Could there be directories that contain a lot of other files besides the ones the model will use?
Yeah, I used to be really worried about including extra data that shouldn't be there, and I feel fairly confident that most users with a `directory` input created the directory for this purpose while assembling their inputs and I think will be less likely to include unrelated files.

In the SWY ET0 case, we don't currently have an `ARGS_SPEC`-defined way to pick out which files are supposed to be there and which aren't since the pattern matching happens within `execute`.  I suppose this would be a reasonable case for the per-model function there and I'd just need to think a bit more about how this might work.

So I think I'd like to leave it alone for now, but I've created https://github.com/natcap/invest/issues/812 for this because the problem is an interesting one and extends to the various sample datasets that we ship with InVEST and might not want to include in the datastacks.

Reply:
Sounds good to me